### PR TITLE
Implement BatchNorm1d

### PR DESCRIPTION
**Pull request by kakao-steve-ai**
## 🙏 Describe the pull request

2D 인풋에 대한 BatchNorm 함수 구현했습니다.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

**Comment by daemyung** [comment]
@ansteve 역전파까지 구현하고 반영하죠.

File: benchmarks/benchmark_batch_norm.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        args={'num_vec': 20},
+        ylabel='milliseconds',
+        x_log=True
+    )
+)
+def bench_batch_norm_forward(num_vec, vec_sz, provider):
+    inp = torch.randn(num_vec, vec_sz, device='cuda')
+
+    if provider == 'torch':
+        return triton.testing.do_bench(lambda: torch.nn.functional.batch_norm(inp, None, None, training=True))
```
Comment (daemyung) [inline benchmarks/benchmark_batch_norm.py:27]: `training`이 `True`여야하는 이유가 있나요?

**Comment by kakao-steve-ai** [comment]
> @ansteve 역전파까지 구현하고 반영하죠.

네 알겠습니다.

File: tests/test_batch_norm.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import trident
+from tests import util
+
+
+@pytest.mark.parametrize('num_vec, vec_sz', [(3, 16), (7, 30)])
+def test_function(num_vec, vec_sz, dtype, device):
+    inp = torch.randn(num_vec, vec_sz, dtype=dtype, device=device)
+    assert util.equal(
+        torch.nn.functional.batch_norm(inp, None, None, training=True),
```
Comment (daemyung) [inline tests/test_batch_norm.py]: `torch.nn.functional.batch_norm(inp, None, None, training=True), trident.function.batch_norm(inp)`를 한 줄로 작성하면 PEP를 위배하나요?

File: trident/module.py [hunk 3]
```diff
@@ -59,6 +59,46 @@ def __view(x):
...
+class BatchNorm1d(torch.nn.Module):
+    def __init__(self, num_features, eps=1e-05, affine=True, device=None, dtype=None):
+        """
+        Applies Batch Normalization over a 2D or 3D inputs.
+
+        Args:
+            num_features: number of features
+            eps: a value added to the denominator for numerical stability
+            affine: a boolean value that when set to True, this module has learnable affine parameters.
+            device: the desired device of returned tensor
```
Comment (daemyung) [inline trident/module.py]: `device`와 `dtype`은 따로 주석을 달지 않으셔도 됩니다.

File: trident/kernel/batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                bs: triton.language.constexpr):
+        pid = triton.language.program_id(0)
+        blk = triton.language.arange(0, bs)
+        inp_blk = blk * vec_sz + pid
+        msk = blk < bt_sz
+
+        inp = triton.language.load(inp_ptr + inp_blk, msk, 0)
+        mean = language.sum(inp) / bt_sz
+        var = language.var(msk, inp, bt_sz, mean, correction=0)
+        std = triton.language.sqrt(var + eps)
```
Comment (daemyung) [inline trident/kernel/batch_norm.py]: `language.std`를 사용하시면 더 좋습니다.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: benchmarks/benchmark_batch_norm.py (hunk 1 above)
Comment (kakao-steve-ai) [reply benchmarks/benchmark_batch_norm.py:27]: `training=False` 인 경우에, `running_mean`, `running_std`를 넣어주어야하는데 어떤 값을 넣어주어야 할지 몰라서 이렇게 했습니다. backward 쪽 구현하면서 좀 더 파악해보겠습니다.

Comment (daemyung) [reply benchmarks/benchmark_batch_norm.py:27]: 네~ 급하게 작업하지 않으셔도 됩니다. 완성도가 중요하니 천천히 잘 구현해 주세요.

File: tests/test_batch_norm.py (hunk 2 above)
Comment (kakao-steve-ai) [reply tests/test_batch_norm.py]: 느낌상 긴 것 같아 줄 바꿈했는데 잘 모르겠네요.  https://github.com/psf/black 같은 파이썬 포맷터를 사용하는거 어떨까요? 전에 사용해봤는데 알아서 한줄로 바꿔주거나, 줄바꿈 해주거나 해서 편하더라고요.

File: trident/module.py (hunk 3 above)
Comment (kakao-steve-ai) [reply trident/module.py]: 네 수정하겠습니다.

File: trident/kernel/batch_norm.py (hunk 4 above)
Comment (kakao-steve-ai) [reply trident/kernel/batch_norm.py]: 네 수정하겠습니다.

File: tests/test_batch_norm.py (hunk 2 above)
Comment (daemyung) [reply tests/test_batch_norm.py]: 좋습니다. 저는 PyCharm의 [Reformat Code](https://www.jetbrains.com/pycharm/guide/tips/reformat-code) 기능을 사용하고 있습니다.

File: benchmarks/benchmark_batch_norm.py (hunk 1 above)
Comment (kakao-steve-ai) [reply benchmarks/benchmark_batch_norm.py:27]: torch 쪽 테스트를 해보니까 `training=False`인 경우에는 `running_mean`, `running_var`을 인자로 받아서 이것을 mean, std 계산에 사용합니다. 이전에는 무조건 커널 내부에서 mean, std 계산하도록 했는데, 이 부분 torch 쪽이랑 작동 맞춰주기 위해서 수정하였습니다.

File: benchmarks/benchmark_batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import triton
+
+import trident
+
+
+@triton.testing.perf_report(
```
Comment (daemyung) [inline benchmarks/benchmark_batch_norm.py]: `util.report` 사용해 주세요.

File: benchmarks/benchmarker.py [hunk 6]
```diff
@@ -50,6 +52,8 @@ def print_scenarios():
 def run_benchmarks(args):
     if args.scenario == 'adaptive-avg-pool2d':
         benchmark_adaptive_avg_pool2d.run_benchmarks(args.mode, args.show_plots)
+    elif args.scenario == 'batch-norm':
```
Comment (daemyung) [inline benchmarks/benchmarker.py:55]: `elif not args.scenario:`에도 `benchmark_batch_norm.run_benchmarks(args.mode, args.show_plots)` 추가해 주세요

File: benchmarks/benchmark_batch_norm.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    if provider == 'torch':
+        return triton.testing.do_bench(lambda: torch.nn.functional.batch_norm(inp, None, None, training=True))
+    else:
+        return triton.testing.do_bench(lambda: trident.function.batch_norm(inp, training=True))
+
+
+def run_benchmarks(mode, show_plots):
+    if mode == 'forward':
+        bench_batch_norm_forward.run(print_data=True, show_plots=show_plots)
+    elif mode == 'backward':
```
Comment (daemyung) [inline benchmarks/benchmark_batch_norm.py:50]: 역전파도 구현하셨으니 역전파 벤치마크도 구현해주세요.

File: trident/function/function.py [hunk 8]
```diff
@@ -24,6 +24,19 @@ def adaptive_avg_pool2d(input, output_size):
...
+    """
+    Applies Batch Normalization for last certain number of dimensions.
+
+    See BatchNorm for details.
+    """
+    if training:
+        running_mean, running_var = None, None
+    else:
+        assert running_mean is not None and running_var is not None
+    return operation.BatchNorm.apply(input, None, None, eps, None, None, running_mean, running_var)
```
Comment (daemyung) [inline trident/function/function.py]: `return` 전에 줄바꿈 있으면 더 좋을 것 같아요.

File: trident/kernel/batch_norm.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+        grad_out = triton.language.load(grad_out_ptr + inp_blk, msk, 0)
+        inp = triton.language.load(inp_ptr + inp_blk, msk, 0)
+        mean = triton.language.load(mean_ptr + pid)
+        std = triton.language.load(std_ptr + pid)
+        wgt = triton.language.load(wgt_ptr + pid) if wgt_ptr is not None else 1
+
+        xc = inp - mean
+        xn = xc / std
+        dxn = wgt * grad_out
```
Comment (daemyung) [inline trident/kernel/batch_norm.py]: `d` 대신에 `grad_`로 바꿔주세요. 저희 코드에서 미분값을 표현할때 `grad`를 사용합니다.

File: trident/module.py [hunk 10]
```diff
@@ -59,6 +59,62 @@ def __view(x):
...
+                and initializes the statistics as None.
+        """
+        super().__init__()
+
+        self.eps = eps
+        self.momentum = momentum
+        self.track_running_stats = track_running_stats
+        self.device = device
+        self.dtype = dtype
+        if affine:
```
Comment (daemyung) [inline trident/module.py:85]: `if-else` 시작과 후에 줄바꿈을 하는게 더 깔끔할 것 같아요.

File: trident/module.py [hunk 11]
```diff
@@ -59,6 +59,62 @@ def __view(x):
...
+            self.weight = torch.nn.Parameter(torch.empty(num_features, device=device, dtype=dtype).fill_(1))
+            self.bias = torch.nn.Parameter(torch.zeros(num_features, device=device, dtype=dtype))
+        else:
+            self.weight = None
```
Comment (daemyung) [inline trident/module.py]: `None`의 경우 `self.register_parameter('weight', None)`로 초기화를 해야하더라고요.

File: trident/module.py [hunk 12]
```diff
@@ -59,6 +59,62 @@ def __view(x):
...
+
+        Returns:
+            an output with the same dimension and shape as an input
+        """
+        assert input.dim() == 2
+        assert input.shape[0] > 1
+
+        mean = torch.empty(input.shape[1], device=self.device, dtype=self.dtype)
+        std = torch.empty(input.shape[1], device=self.device, dtype=self.dtype)
+        if self.track_running_stats:
```
Comment (daemyung) [inline trident/module.py:112]: `if-else` 시작과 후에 줄바꿈을 하는게 더 깔끔할 것 같아요.

**Review by daemyung — CHANGES_REQUESTED** [review]

**Comment by daemyung** [comment]
@ansteve 스타일 관련된 커멘트는 무시하세요. `black`을 사용할 예정입니다.

**Comment by kakao-steve-ai** [comment]
> @ansteve 스타일 관련된 커멘트는 무시하세요. `black`을 사용할 예정입니다.

네 알겠습니다~

**Comment by kakao-steve-ai** [comment]
@daemyung 코멘트 주신 것들 수정하였습니다.

**Review by daemyung — APPROVED** [review]
LGTM
